source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -108,6 +110,13 @@ Config::Config(const envoy::config::filter::network::tcp_proxy::v2::TcpProxy& co
 }
 
 const std::string& Config::getRegularRouteFromEntries(Network::Connection& connection) {
+  // First check if the per-connection state to see if we need to route to a pre-selected cluster
+  if (connection.perConnectionState().hasData<PerConnectionCluster>(PerConnectionCluster::KEY)) {
```

Comment:
not directly with this PR, but seems this pattern (hasData, then getData) would be common, is there any way to avoid double lookup?

Reply:
I had this thought too, but was thinking we should fix it in a separate PR.  Probably just make the accessor return a const pointer instead of const ref, so that it can be nullptr when it isn't found. And maybe throw if the item is found, but has an incompatible type (I think that always indicates a programmer error).